---
permalink: /science/
layout: single
title: "Science"
toc: true
toc_label: "On this page"
toc_icon: "cog"
---

As an academic in training, I aim to contribute to future food security by working on a highly productive plant and novel crop: _Azolla_.
In our _Azolla_ research group, we benefit from each others' specialities; 
I support my colleagues with a computational infrastructure, technical issues and intricacies of data analysis for example. 
Additionally, I take joy in initiating a lot of our lab's outreach activities. 
{: .text-justify}


# The Azolla metagenome
In my PhD project, I work on the metagenome (genomes of bacteria associated with) the amazing fern _Azolla_. 
_Azolla_ is an aquatic floating fern, known for its high growth rate and protein content.
Both these propperties, are a direct consequence of the special symbiosis _Azolla_ has with a nitrogen fixing cyanobacterium inside its leaves.
When domesticated propperly, the _Azolla_ symbiosis can be a sustainable protein source grown in temperate regions,
thereby reducing the need for imported protein grown in tropical regions.
{: .text-justify}

## Foul play in the pocket
In the Foul play in the pocket paper, my colleagues and I stumbled upon not one, but multiple bacteria residing inside the leaves of _Azolla_.
Whilst the known symbiont is by far the most abundant, multiple other microbes seem to be present systematically too. 
Similar bacteria were found to be present in other species of the _Azolla_ genus.
These groups of bacteria are known to also fix nitrogen in some symbiosis, but genomic and nitrogen isotope experiments pointed out that this is not the case here in _Azolla_.
[Is there foul play in the leaf pocket? The metagenome of floating fern Azolla reveals endophytes that do not fix N2 but may denitrify](https://doi.org/10.1111/nph.14843)
{: .text-justify}

![lowly abundant microbes in _Azolla_ species](https://nph.onlinelibrary.wiley.com/cms/asset/a5839007-9b9b-4a24-9145-6407ee51562f/nph14843-fig-0002-m.jpg)
_See our publication linked above for more details_
{: .text-justify}

## Partners & Passengers
More details later...


# Phylogeny workflow
My projects primary interest is the microbial community associated with _Azolla_, yet I also work with colleages to understand more about the physiology of the fern itself.
One way I help, is infering the evolutionary history of certain proteins of our interest.
While learning to reliably infer such a phylogeny, I decided to document my workflow for others, and published this on [Github](https://github.com/lauralwd/lauras_phylogeny_wf).
With this workflow, students, colleagues, or anyone else, can use my methods to infer their phylogeny.
This work was achieved only with generous help and advice from the Berend Snel group.
Some examples of how I used this workflow myself, are also on github. 
These are examples of how I learn to practice Open Science, documenting all my actions, analyses, log files, notes and mistakes using jupyter notebooks and version control software Git.
{: .text-justify}

 - LAR phylogeny [GitHub](https://github.com/lauralwd/LAR_phylogeny_gungor-et-al-2020/tree/v1.00.00) [Zenodo](https://doi.org/10.5281/zenodo.3959057)
 - MIKC phylogeny [GitHub](https://github.com/lauralwd/MIKC_tree) [Zenodo](https://doi.org/10.5281/zenodo.4564374)
 - R2R3 MYB phylogeny [GitHub](https://github.com/lauralwd/azolla_MYBs) [Zenodo](https://doi.org/10.5281/zenodo.4564441)




